cumulus/pallets/parachain-system/src/block_weight/transaction_extension.rs

```diff
@@ -0,0 +1,489 @@
+					let extrinsic_class_weight = block_weight.get(info.class);
+
+					if extrinsic_class_weight.any_gt(target_weight) {
+						log::trace!(
+							target: LOG_TARGET,
+							"Extrinsic class weight {extrinsic_class_weight:?} above target weight {target_weight:?}, enabling `FullCore` mode."
+						);
+
+						*weight_mode = Some(BlockWeightMode::FullCore);
+
+						// Inform the node that this block uses the full core.
+						frame_system::Pallet::<Config>::deposit_log(
+							CumulusDigestItem::UseFullCore.to_digest_item(),
+						);
+					} else {
+						log::trace!(
+							target: LOG_TARGET,
+							"Extrinsic class weight {extrinsic_class_weight:?} not above target \
+							weight {target_weight:?}, going back to `FractionOfCore` mode."
+						);
+
+						*weight_mode =
+							Some(BlockWeightMode::FractionOfCore { first_transaction_index });
+					}
+
+					// We run into the worst case, so no refund :)
+					Weight::zero()
+				},
+			}
+		})
+	}
+}
+
+impl<
+		Config,
+		Inner,
+		TargetBlockRate,
+		const MAX_TRANSACTION_TO_CONSIDER: u32,
+		const ALLOW_NORMAL: bool,
+	> From<Inner>
+	for DynamicMaxBlockWeight<
+		Config,
+		Inner,
+		TargetBlockRate,
+		MAX_TRANSACTION_TO_CONSIDER,
+		ALLOW_NORMAL,
+	>
+{
+	fn from(s: Inner) -> Self {
+		Self::new(s)
+	}
+}
+
+impl<
+		Config,
+		Inner: core::fmt::Debug,
+		TargetBlockRate,
+		const MAX_TRANSACTION_TO_CONSIDER: u32,
+		const ALLOW_NORMAL: bool,
+	> core::fmt::Debug
+	for DynamicMaxBlockWeight<
+		Config,
+		Inner,
+		TargetBlockRate,
+		MAX_TRANSACTION_TO_CONSIDER,
+		ALLOW_NORMAL,
+	>
+{
+	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+		write!(f, "DynamicMaxBlockWeight<{:?}>", self.0)
+	}
+}
+
+impl<
+		Config: crate::Config + Send + Sync,
+		Inner: TransactionExtension<Config::RuntimeCall>,
+		TargetBlockRate: Get<u32> + Send + Sync + 'static,
+		const MAX_TRANSACTION_TO_CONSIDER: u32,
+		const ALLOW_NORMAL: bool,
+	> TransactionExtension<Config::RuntimeCall>
+	for DynamicMaxBlockWeight<
+		Config,
+		Inner,
+		TargetBlockRate,
+		MAX_TRANSACTION_TO_CONSIDER,
+		ALLOW_NORMAL,
+	>
+where
+	Config::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
+{
+	const IDENTIFIER: &'static str = "DynamicMaxBlockWeight<Use `metadata()`!>";
+
+	type Implicit = Inner::Implicit;
+
+	type Val = Inner::Val;
+
+	type Pre = Inner::Pre;
+
+	fn implicit(&self) -> Result<Self::Implicit, TransactionValidityError> {
+		self.0.implicit()
+	}
+
+	fn metadata() -> Vec<sp_runtime::traits::TransactionExtensionMetadata> {
+		let mut inner = Inner::metadata();
+		inner.push(sp_runtime::traits::TransactionExtensionMetadata {
+			identifier: "DynamicMaxBlockWeight",
+			ty: scale_info::meta_type::<()>(),
+			implicit: scale_info::meta_type::<()>(),
+		});
+		inner
+	}
+
+	fn weight(&self, _: &Config::RuntimeCall) -> Weight {
+		Config::WeightInfo::block_weight_tx_extension_max_weight()
+	}
```

Comment:
we need to add the inner weight as well.
```suggestion
	fn weight(&self, call: &Config::RuntimeCall) -> Weight {
		Config::WeightInfo::block_weight_tx_extension_max_weight().saturating_add(self.0.weight(call))
	}
```